### PR TITLE
Support imported functions

### PR DESCRIPTION
**Pull request by tessi**
This PR adds support for importing elixir functions into WASM modules.
It wraps elixir function references and makes them available to WASM modules. A call follows roughly this roundtrip:

1. `invoke_wasm_function (elixir)`
2. `run function in wasm/rust`
3. `trigger callback in a rust wrapper`
4. `the wrapper forwards the callback execution back to elixir-land`
5. `elixir runs the callback and gives the results back`
6. `the wasm function continues with the callback results`

See test/wasmex_test.exs for examples

```elixir
imports = %{
  env: %{
    sum3: {:fn, [:i32, :i32, :i32], [:i32], fn (_context, a, b, c) -> a + b + c end},
  }
}
instance = start_supervised!({Wasmex, %{bytes: @import_test_bytes, imports: imports}})

{:ok, [6]} = Wasmex.call_function(instance, "use_the_imported_sum_fn", [1, 2, 3])
```

Note: It currently needs `wasmex` from the current git-master - I expect them to release a new version until I'm done cleaning this PR up.

TODOs

* [x] build proper signatures. See #12 
* [x] think about a way to pass the modules memory into the callback. See #13 
* [x] think how to support other types of imports (whole instances, tables, globals) - the API should allow us to add these in a later time
* [x] add support for atom-keys in elixir-imports object
* [x] ~theoretically wasm allows many return values, but elixir only has one return value for a function. think about a way to allow multiple return values~ (not changing things here for now, our internal API already handles returning an array of return values. If we ever needs this, we can do it later)
* [x] elixir callbacks should receive a tuple of `{context, params}` instead of just `params` provide a stable API. We plan to inject context (e.g. memory) into callbacks in the future
* [x] add info to the README
* [x] add to the changelog
* [x] make the wasmer dependency non-local again (and remove the git submodule)
* [x] add method/module documentation
* [x] chaining/recursion: what if a callback calls into another WASM function 🤯 (probably not possible due to the GenServer calling itself) -> yep, also a special case we can ignore for now
* [x] ~think about adding helper functions to make constructing function imports easy~ not in this PR at least. We may add some later, when people need it
* add more tests
  * test non-matching signatures from the given spec and the elixir callback (both params and return values)
    * [x] wrong datatypes got chosen as functions params
    * [x] wrong datatypes got chosen as functions returns
    * [x] too many/few params are given in the signature
    * [x] too many/few returns are given in the signature
  * [x] the elixir callback throws an exception (=> never goes back into Rust to free the lock waiting on a callback return) => find out how to handle this situation

### CODE REVIEW

File: native/wasmex/Cargo.toml [hunk 1]
```diff
@@ -17,5 +17,5 @@ crate-type = ["dylib"]
 [dependencies]
 rustler = "0.21.0"
 lazy_static = "1.0"
-wasmer-runtime = "0.16.2"
-wasmer-runtime-core = "0.16.2"
+wasmer-runtime = {path = "../../../wasmer/lib/runtime", version = "0.16.2"}
+wasmer-runtime-core = {path = "../../../wasmer/lib/runtime-core", version = "0.16.2", features = ["dynamicfunc-fat-closures"]}
```
Comment (tessi) [inline native/wasmex/Cargo.toml:21]: This needs to be reverted before merge/release. Specifying a path is not working when uploading to crates.io and the old release version (0.16.2) does provide support for the `dynamicfunc-fat-closures` feature - it's on `master` only for now.

**Comment by tessi** [comment]
@myobie knowing that you want to work on implementing proper signatures, I realized I started work on it (but got distracted by life.

Maybe https://github.com/tessi/wasmex/commit/90cfe9aa203ea69aeb91152c0aeba7e8f6d585c8 can give you a pointer on where to start looking.

The commit does the following so far:

* it extends the callback-token to also include a "return signature" so that elixir params can be converted back to WASM land
* It works on digesting the given elixir signature and make it an `Vec<String>` for easier handling.
* improves wording (`results` vs. `returns` to be consistent with `FuncSig`)

Feel free to start from that commit or invent your own :)
Next steps could be:

* take the `Vec<String>` for both `params` ([here](https://github.com/tessi/wasmex/commit/90cfe9aa203ea69aeb91152c0aeba7e8f6d585c8#diff-affd2b0fae2d52f64d6570fd6301ae02R58)) and `returns` ([here](https://github.com/tessi/wasmex/commit/90cfe9aa203ea69aeb91152c0aeba7e8f6d585c8#diff-affd2b0fae2d52f64d6570fd6301ae02R119)) and replace the fake signature with a proper one
* extend test/wasmex_test.exs (probably somewhere at the bottom of the file) with a use case more specific to yours (e.g. that other function signature you wanted). Getting that test green could be our next goal.
* totally optional: start documenting things - if you read a weird method and make sense of it, write down your thoughts (or confusion) so that we can either document or improve it.

Meta Note: Please note that this is my biggest Rust project so far - I'm still learning many things here. Also, most of the code is not as clean as I like it yet. Feel welcome to either implement or suggest/discuss improvements.

**Comment by tessi** [comment]
I *think* this PR is ready feature-wise. What is missing are more tests and documentation.

**Comment by tessi** [comment]
* wrote some more docs (but there are still some missing)
* added a test for an elixir-callback throwing an exception and handling that with a rust-side panic in the imported-function-wrapper
* removed the git submodule of wasmex and instead point to master (should have seen that option in the beginning, turns out a submodule/local path wasn't necessary at all)
* sidetask: optimized CI build time from ~13 minutes to ~3 minutes
* ran the elixir linting and fixed upcoming suggestions
* added rust-clippy to our CI checks and applied it's suggestions (it made some of our rust code better 🎉 )

**Comment by tessi** [comment]
@myobie I just wrote the missing tests (or do you think I missed some?)

I would do another round of documenting things later and then merge this PR.

File: native/wasmex/src/instance.rs [hunk 2]
```diff
@@ -1,27 +1,39 @@
...
 pub struct InstanceResource {
     pub instance: Mutex<runtime::Instance>,
 }
 
+// creates a new instance from the given WASM bytes
+// expects the following elixir params
+//
+// * bytes (binary): the bytes of the WASM module
+// * imports (map): a map defining eventual instance imports, may be empty if there are none.
+//   structure: %{namespace_name: %{import_name: {TODO: signature}}}
```
Comment (myobie) [inline native/wasmex/src/instance.rs:25]: We should probably replace this TODO. I'll add this to my todo list.

**Comment by myobie** [comment]
@tessi OMG sorry I was late getting back to you, this PR looks great to me! The docs inline in the elixir are great and I really appreciate the additional tests like https://github.com/tessi/wasmex/pull/9/files#diff-3ea07b28dd9682257c33fd0f4f7a0706R38 and https://github.com/tessi/wasmex/pull/9/files#diff-3ea07b28dd9682257c33fd0f4f7a0706R78

Is it going to be a problem that the submodule change was included? I think it's fine, but wanted to call it out to make sure you didn't accidentally forget.

File: native/wasmex/src/instance.rs (hunk 2 above)
Comment (tessi) [reply native/wasmex/src/instance.rs:25]: ohh, good catch 🌻 thanks!

**Comment by tessi** [comment]
please don't be sorry! You helped me a lot in this PR - not only in source code, but also in ideas and motivation. I'm sure it would still be sitting on my unfinished commit if you weren't here. 💜 

> Is it going to be a problem that the submodule change was included?

ha, you have good eyes, I thought I removed all traces of the submodule but didn't see that one. will remove it on master in a second. It shouldn't be a problem though.

**Pull request merged**